### DISCUSSION TITLE
Reusable Query parameters (and perhaps Path)

### DISCUSSION THREAD

**Discussion by jacobsvante** [fastapi/fastapi#9051] [Questions]
### Is your feature request related to a problem

Well, maybe.

In one of my fastapi-based apps there's a lot of query parameters supported by multiple GET endpoints, basically for adding filters for each field on a pydantic model (with multiple operators per field). They are generated dynamically with some `inspect` magic.

My problem is that the OpenAPI spec becomes really heavy. 2MB+ easily. And with each added endpoint that reuses the same huge list of parameters it adds another few 100kb.

I haven't verified it yet but I think this is what leads the app to take 5+ seconds to start.

### The solution you would like

I'm proposing that all parameters are defined under `components.parameters` in the OpenAPI spec and then referenced in `parameters` for each endpoint.

Of course one would have to ensure that name collisions do not occur for query parameters with the same name under different endpoints.

I think something like a registry concept would make sense here.

Code example:

```python

from fastapi import FastAPI, Query, Parameters
from . import db

app = FastAPI()
parameters = Parameters()

@app.get("/items")
async def get_items(
    # The old way still works, this will be explicity defined under the endpoint's `parameters` object in the OpenAPI spec
    name: str = Query(...),
    # Using lru_cache this would return the same param, which would be defined under `components.parameters` in the OpenAPI spec
    per_page: str = parameters.Query(..., description="Results per page")
):
    return await db.get_items(name=name, limit=per_page)
```

Or a more explicit alternative:

```python

from fastapi import FastAPI, Parameters, Query
from . import db

app = FastAPI()
parameters = Parameters()
per_page_query = parameters.Query(str, ..., description="Results per page")  # NOTE: Additional type argument

@app.get("/items")
async def get_items(
    name: str = Query(...),
    per_page: str = per_page_query
):
    return await db.get_items(name=name, limit=per_page)
```

The concept should work nicely with path parameters as well:

```python
import uuid
from fastapi import FastAPI, Parameters, Query
from . import db

app = FastAPI()
parameters = Parameters()
item_id_path = parameters.Path(uuid.UUID, ..., description="Item ID")

@app.get("/items/{item_id}")
async def get_item(
    item_id: uuid.UUID = item_id_path
):
    return await db.get_item(item_id)
```

### Describe alternatives you've considered

1. Not adding the parameter definitions to the API and instead manually iterating over them via `starlette.Request.query_params`. Disadvantages: No for-free validation on query parameters, no automatic API documentation.
2. Using a special query parameter syntax. E.g. `?filter=field:operator:value` instead of `?filter:operator=value`. Disadvantages: Additional typing for the API user, less clear API documentation.

**Comment by thatguysimon** [comment]
Hi @jacobsvante,
What did you end up doing to solve this?

**Comment by soundstripe** [comment]
I use this pattern (note -- must be defined BEFORE/ABOVE where it is used in a path operation):

```
@lru_cache  # <--- optional, I just mark known "static" functions this way
def QueryIncludeMeta():
    """Helper to reuse query parameter documentation, example, etc"""
    return Query(
        default=False,
        description="If true, object meta-information will be returned along with the object",
    )


@app.get('/something')
def get_something(include_meta: bool = QueryIncludeMeta()):
    return 'something'
```

it also works to just use a global but I don't like it as much as it doesn't "look" like a FastAPI tool.

```
# Helper to reuse query parameter documentation, example, etc
QueryIncludeMeta = Query(
      default=False,
      description="If true, object meta-information will be returned along with the object",
 )


@app.get('/something')
def get_something(include_meta: bool = QueryIncludeMeta):
    return 'something'
```

**Comment by de-code** [comment]
Did you consider using [Dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/)?
